src/CodeGen_Hexagon.cpp

```diff
@@ -326,350 +313,386 @@ void CodeGen_Hexagon::compile_func(const LoweredFunc &f,
-
-        // Bitwise operators
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vand),  u8v1,  "and.vb.vb",  {u8v1,  u8v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vand),  u16v1, "and.vh.vh",  {u16v1, u16v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vand),  u32v1, "and.vw.vw",  {u32v1, u32v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vor),   u8v1,  "or.vb.vb",   {u8v1,  u8v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vor),   u16v1, "or.vh.vh",   {u16v1, u16v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vor),   u32v1, "or.vw.vw",   {u32v1, u32v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vxor),  u8v1,  "xor.vb.vb",  {u8v1,  u8v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vxor),  u16v1, "xor.vh.vh",  {u16v1, u16v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vxor),  u32v1, "xor.vw.vw",  {u32v1, u32v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vnot),  u8v1,  "not.vb",     {u8v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vnot),  u16v1, "not.vh",     {u16v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vnot),  u32v1, "not.vw",     {u32v1} },
-
-        // Broadcasts
-        { IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatb), u8v1,   "splat_v62.b", {u8}  },   // v62 or later
-        { IPICK(is_128B, Intrinsic::hexagon_V6_lvsplath), u16v1,  "splat_v62.h", {u16} },   // v62 or later
-        { IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw), u32v1,  "splat.w", {u32} },
-
-        // Bit counting
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vcl0h), u16v1, "clz.vh", {u16v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vcl0w), u32v1, "clz.vw", {u32v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vnormamth), u16v1, "cls.vh", {u16v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vnormamtw), u32v1, "cls.vw", {u32v1} },
-        { IPICK(is_128B, Intrinsic::hexagon_V6_vpopcounth), u16v1, "popcount.vh", {u16v1} },
-        // TODO: If we need it, we could implement a popcountw in the
-        // runtime module that uses popcounth, and horizontally add
-        // each pair of lanes.
-    };
-    // TODO: Many variants of the above functions are missing. They
-    // need to be implemented in the runtime module, or via
-    // fall-through to CodeGen_LLVM.
-    for (HvxIntrinsic &i : intrinsic_wrappers) {
-        define_hvx_intrinsic(i.id, i.ret_type, i.name, i.arg_types, i.flags);
+    const bool is_128B = target.has_feature(Halide::Target::HVX_128);
+    vector<Type> arg_types;
+    for (const HvxIntrinsic &i : intrinsic_wrappers) {
+        Intrinsic::ID id = i.ids.get(is_128B);
+        internal_assert(id != Intrinsic::not_intrinsic);
+        // Get the real intrinsic.
+        llvm::Function *intrin = Intrinsic::getDeclaration(module.get(), id);
+        arg_types.clear();
+        for (const auto &a : i.arg_types) {
+            if (a.bits == 0) {
+                break;
+            }
+            const int lanes_actual = a.lanes > 1 ?
```

Comment:
Doesn't the return type need this treatment too?

Reply:
Good catch!